### PR TITLE
Lavaland Mining Rework #1: What do you mean I have to mine??

### PR DESCRIPTION
**Pull request by ToasterBiome**
# ALL SUBJECT TO CHANGE

## READ THIS FOR REASONING BEHIND THE PR

The purpose of this PR is to make the mining portion of mining more interesting. For years mining has gone virtually unchanged except for a new plasma scattergun. While we've had loot changes and weapons, mining hasn't been given the love it deserves.

Yes, we can introduce new loot/ruins/bosses to enhance the experience too, but an important thing to point out is that all of this is still possible with this PR. We can have new minerals, hardness, loot, ruins, bosses, etc. It's not one or the other. It's just more stuff to do.

With this PR I am hoping to introduce new and interesting ways to mine while not stepping on anyone's toes. Some people believe hardness to be a backwards step as it just makes it more tedious, but I think it adds more complexity. Would you enjoy mining in Minecraft or Terraria if you can instantly break every block and collect every item right at the start? Then why here? We do only have two hour rounds, which is far flung from the hours you can spend in the other games, which is why I want to make it meaningful that you are getting these upgrades and more importantly that they are explicably adding to the experience, not taking away.

## THE REST OF IT

Introduces mineral hardness. The normal grey minerals have a hardness of 1. Darker **Hardened Basalt** has a hardness of 2.

Every time you mine, be it a kinetic accelerator, drakeling, plasma cutter, mech, hierophant club, etc. it removes 1 hardness from the mineral. If it has 0 hardness, it breaks and your ores drop. Simple. if it's still left after hitting, it will have a crack in it (a la Minecraft) which will then take another swing/shot/stomp/whatever to break it. Plasma cutters will pierce it and do a line of cracks. 

Hardened Basalt spawns more ores in it and more minerals drop per mineral destroyed.

The Kinetic Accelerator has a **Hardness Upgrade** that takes 10% mod capacity and increases the power per shot by one. This means it will break hardened basalt in one shot. It is 1200 mining points from the Mining Equipment Vendor

Introduces **Mining Charges**. They are **MINERAL ONLY** explosives that breaks a lot in a small area. 5 second timer. Can only be placed on minerals to stop shenanigans. You now get bonus mineral drops from minerals you break with explosives. They are in the Mining Equipment Vendor for 500 mining points for shaft miners to play with. They are only at the top right now so people see them and play with them. They might be made with Gibtonite in the future. This does mean that you can potentially get up to 9 diamonds in one ore if you blow up a hardened basalt diamond ore with Gibtonite/mining charges. Yippee.

kUxvURd4jN
(also I need a sprite)

**Map distribution:**
image
(distribution and shape subject to change because damn this is fucking ugly)

Adds **Plasma Magmite**. A rare ore that spawns exclusively in the hardened basalt. Beware, as it can only be dropped if the mineral it's contained in is destroyed by an explosion (mining charge, gibtonite, etc).

image

image

Take it to the **World Anvil** locaed in front of the necropolis gate. The World Anvil must be charged by using Gibtonite to upgrade your Kinetic Accelerator or Advanced Plasma Cutter. Placing a piece of Plasma Magmite and your desired item to upgrade on the heated anvil and then pressing the anvil will upgrade them into new items:

- **Mega Proto-Kinetic Accelerator**
This functions the same as the normal PKA but has 120% mod capacity (up from 100%)
- **Mega Advanced Plasma Cutter**
This functions the same as an Advanced Plasma Cutter but goes 7 tiles with a bonus range of 7 tiles (up from 5/5)

image

# Wiki Documentation

Everything told in here

# Changelog

:cl:  
rscadd: Adds rock hardness
rscadd: Add PKA mod to increase what hardness rocks you can destroy
rscadd: Adds plasma magmite
rscadd: Adds world anvil to necropolis gate
tweak: adds hard rocks to lavaland
tweak: just read the PR
/:cl:

### CODE REVIEW

**Comment by patpol4** [comment]
I think this is good? I don't really play miner enough to be able to tell.

**Comment by ToasterBiome** [comment]
> I think this is good? I don't really play miner enough to be able to tell.

Hoping for a testmerge to see if it has positive player reception so people can form their own opinions on it instead of "muh miner change"

**Comment by Marmio64** [comment]
perhaps allow pickaxes to break hard rock INCREDIBLY slowly, so that you cant get stuck after escape podding or getting spawned in hard rock as ashwalkers, hermits, etc

**Comment by SomeguyManperson** [comment]
How does this make people want to mine more
It adds special rocks with higher mineral density but the equipment required to break them is either a 1200 point mod or a plasmacutter and there's no real challenge in finding them or digging through them effectively
It seems like the only real difference for miners would be that they need to use twice as meny cutter shots to break through the same number of dense rock as non dense rock

**Comment by ToasterBiome** [comment]
> How does this make people want to mine more It adds special rocks with higher mineral density but the equipment required to break them is either a 1200 point mod or a plasmacutter and there's no real challenge in finding them or digging through them effectively It seems like the only real difference for miners would be that they need to use twice as meny cutter shots to break through the same number of dense rock as non dense rock

These are good points. I do not want to be make it prohibitively expensive because I don't want miners feeling like they're 1) required to get this upgrade and 2) complain that I ruined mining because there is some slightly darker rocks. Moreso, I don't want to be adding something that would make it so you CAN'T mine them at all (I go back on this point slightly later)

I've been thinking recently about introducing mining charges that break these harder rocks and are crafted with gibtonite - obviously they would not be effective on station (for grief). They would be pretty much near useless for fighting fauna, so as the crusher is geared towards fighting - these might be geared more towards mining, and a fun tool to play with while also getting more ores (who doesn't want to blow shit up?).

Another option is making these harder areas much smaller and denser - but are unbreakable except for a special pneumatic pickaxe or other tool item. It would reward you with many riches if you decide to sling one with you, giving you good reason to bring the item and also seek them out.

Of course, a third option is a new ore that's this basalt region only. Despite Yogs' unique ore being near useless, I think with more uses outside very specific fusion (you know which ore i'm talking about) it will make people seek it out.

I am not trying to make Lavaland mining so soulcrushing that it isn't fun. I am trying to think of things that could spice up the mining part in more complex ways as it is our only mining level, and while fighting megafauna is fun - I wish there was as much variety in the actual mining part as the "fighting fauna" part. It has pretty much stayed unchanged from asteroid mining, except for possibly plasma cutters. To make it more dynamic where you have more options from just "get the plasma cutter and run in a circle", which I understand this PR in it's current state is perpetuating just adding time to the meta with plasma cutters taking longer - which is why I am asking for other ways you would believe hard basalt to be used in an interesting way.

Feel free to hit me up in DMs if you want to discuss this further.

**Comment by ilikesaying** [comment]
can you integrate the hardness upgrades to the exosuit mining equipment? they would benefit from this due to the heavy duty nature of exosuits.

**Review by nesisknesis — APPROVED** [review]

**Review by nesisknesis — APPROVED** [review]

**Comment by githubuser4141** [comment]
How does this affect mining mech balance? Currently if you're good with strafe/your hands you can mine straight through common hazards (goliath tentacles/legion skulls/lava) so while a Firefighter is slower then a full-health person in a straight line the way it picks up ores in a 3x1 radius makes it a good deal quicker IF you're robust with mechs. So. I hope this won't nerf mech mining into the ground.

**Review by patpol4 — APPROVED** [review]
the map is fine

**Comment by kooarbiter** [comment]
maybe leave some charges in the gulag so we can recreate the powder gangers

**Comment by ToasterBiome** [comment]
> maybe leave some charges in the gulag so we can recreate the powder gangers

local sec officer gets his legs blown off by gulag prisoners

**Comment by ToasterBiome** [comment]
new shit just dropped

File: code/modules/mining/ores_coins.dm [hunk 1]
```diff
@@ -293,6 +293,14 @@ GLOBAL_LIST_INIT(sand_recipes, list(\
 		return
 	qdel(src)
 
+//plasa magmite functions like an ore in the way it's mined, but we don't want to be redeeming this. Since Gibtonite is stuck with other ores it's okay to put it here.
+/obj/item/magmite
+	name = "plasma magmite"
+	desc = "A chunk of plasma magmite, crystallized deep under lavaland's surface. Its strength seems to fluctuate depending on the distance to the planet."
+	icon = 'icons/obj/mining.dmi'
+	icon_state = "Magmite ore"
```
Comment (Bibby0110) [inline code/modules/mining/ores_coins.dm:301]: no

File: code/modules/mining/lavaland/world_anvil.dm [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+			forge_charges--
+			used_magmite++
+			to_chat(user,"You forge an upgrade to your advanced plasma cutter. It now goes extra distance.")
+	//time to clean up all the magmite we used
+	for(var/obj/item/magmite in placed_objects)
+		if(used_magmite != 0)
+			used_magmite--
+			qdel(magmite)
+	update_icon()
+	if(forge_charges == 0)
```
Comment (Bibby0110) [inline code/modules/mining/lavaland/world_anvil.dm]:
```suggestion
	if(!forge_charges)
```

File: code/modules/mining/lavaland/world_anvil.dm [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+			placed_objects += I
+			RegisterSignal(I, COMSIG_MOVABLE_MOVED, .proc/ItemMoved,TRUE)
+
+/obj/structure/world_anvil/proc/ItemMoved(obj/item/I, atom/OldLoc, Dir, Forced)
+	vis_contents -= I
+	placed_objects -= I
+	UnregisterSignal(I, COMSIG_MOVABLE_MOVED)
+
+/obj/structure/world_anvil/attack_hand(mob/user)
+	if(LAZYLEN(placed_objects) == 0)
```
Comment (Bibby0110) [inline code/modules/mining/lavaland/world_anvil.dm]:
```suggestion
	if(!LAZYLEN(placed_objects))
```

File: code/modules/mining/lavaland/world_anvil.dm [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+		to_chat(user,"You must place a piece of plasma magmite and either a kinetic accelerator or advanced plasma cutter on the anvil!")
+		return ..()
+	if(forge_charges <= 0)
+		to_chat(user,"The anvil is not heated enough to be usable!")
+		return ..()
+	var/magmite_amount = 0
+	var/used_magmite = 0
+	for(var/obj/item/magmite in placed_objects)
+		magmite_amount++
+	if(magmite_amount == 0)
```
Comment (Bibby0110) [inline code/modules/mining/lavaland/world_anvil.dm]:
```suggestion
	if(!magmite_amount)
```

File: code/modules/mining/lavaland/world_anvil.dm [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+		to_chat(user,"The anvil does not have any plasma magmite on it!")
+		return ..()
+	for(var/obj/item/I in placed_objects)
+		if(istype(I,/obj/item/gun/energy/kinetic_accelerator) && !istype(I,/obj/item/gun/energy/kinetic_accelerator/crossbow) && forge_charges > 0 && used_magmite < magmite_amount)
```
Comment (Bibby0110) [inline code/modules/mining/lavaland/world_anvil.dm]:
```suggestion
		if(istype(I,/obj/item/gun/energy/kinetic_accelerator) && !istype(I,/obj/item/gun/energy/kinetic_accelerator/crossbow) && forge_charges && used_magmite < magmite_amount)
```

File: code/modules/mining/lavaland/world_anvil.dm [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+			var/obj/item/gun/energy/kinetic_accelerator/gun = I
+			gun.max_mod_capacity = 120 //should make a new gun and replace it, just for testing
```
Comment (Bibby0110) [inline code/modules/mining/lavaland/world_anvil.dm]: what?

File: code/modules/mining/lavaland/world_anvil.dm [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+			forge_charges--
+			used_magmite++
+			to_chat(user,"You forge an upgrade to your kinetic accelerator. It now can hold 120% mod capacity.")
+		if(istype(I,/obj/item/gun/energy/plasmacutter/adv) && forge_charges > 0 && used_magmite < magmite_amount)
```
Comment (Bibby0110) [inline code/modules/mining/lavaland/world_anvil.dm]:
```suggestion
		if(istype(I,/obj/item/gun/energy/plasmacutter/adv) && forge_charges && used_magmite < magmite_amount)
```

File: code/modules/mining/lavaland/world_anvil.dm [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+			var/obj/item/gun/energy/plasmacutter/adv/cutter = I
+			cutter.ammo_type = list(/obj/item/ammo_casing/energy/plasma/adv/forge) //should make a new gun and replace it, just for testing
```
Comment (Bibby0110) [inline code/modules/mining/lavaland/world_anvil.dm]: what

File: code/modules/mining/lavaland/world_anvil.dm [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+			forge_charges--
+			used_magmite++
+			to_chat(user,"You forge an upgrade to your advanced plasma cutter. It now goes extra distance.")
+	//time to clean up all the magmite we used
+	for(var/obj/item/magmite in placed_objects)
+		if(used_magmite != 0)
```
Comment (Bibby0110) [inline code/modules/mining/lavaland/world_anvil.dm]:
```suggestion
		if(used_magmite)
```

File: code/modules/mining/lavaland/world_anvil.dm [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+	if(forge_charges > 0)
+		set_light(4,1,LIGHT_COLOR_ORANGE)
+	else
+		set_light(0)
+
+/obj/structure/world_anvil/examine(mob/user)
+	. = ..()
+	. += "It currently has [forge_charges] forge[forge_charges != 1 ? "s" : ""] remaining."
+
+/obj/structure/world_anvil/attackby(obj/item/I, mob/living/user, params)
```
Comment (Bibby0110) [inline code/modules/mining/lavaland/world_anvil.dm:27]: need ..()

File: code/modules/mining/equipment/mining_charges.dm [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+	alert_admins = FALSE
+
+/obj/item/grenade/plastic/miningcharge/Initialize()
+	. = ..()
+	plastic_overlay = mutable_appearance(icon, "[icon_state]_active", ON_EDGED_TURF_LAYER)
+
+/obj/item/grenade/plastic/miningcharge/attack_self(mob/user)
+	if(nadeassembly)
+		nadeassembly.attack_self(user)
+		return
```
Comment (Bibby0110) [inline code/modules/mining/equipment/mining_charges.dm]:
```suggestion
```

File: code/game/turfs/simulated/minerals.dm [hunk 12]
```diff
@@ -720,3 +793,25 @@
 	turf_type = /turf/open/floor/plating/asteroid/snow/ice/icemoon
 	baseturfs = /turf/open/floor/plating/asteroid/snow/ice/icemoon
 	initial_gas_mix = ICEMOON_DEFAULT_ATMOS
+
+/turf/closed/mineral/magmite
+	mineralType = /obj/item/magmite
+	spread = 0
+	scan_state = "rock_Magmite"
+
+/turf/closed/mineral/magmite/gets_drilled(mob/user, triggered_by_explosion = 0)
```
Comment (Bibby0110) [inline code/game/turfs/simulated/minerals.dm]:
```suggestion
/turf/closed/mineral/magmite/gets_drilled(mob/user, triggered_by_explosion = FALSE)
```

File: code/game/turfs/simulated/minerals.dm [hunk 13]
```diff
@@ -720,3 +793,25 @@
 	turf_type = /turf/open/floor/plating/asteroid/snow/ice/icemoon
 	baseturfs = /turf/open/floor/plating/asteroid/snow/ice/icemoon
 	initial_gas_mix = ICEMOON_DEFAULT_ATMOS
+
+/turf/closed/mineral/magmite
+	mineralType = /obj/item/magmite
+	spread = 0
```
Comment (Bibby0110) [inline code/game/turfs/simulated/minerals.dm:799]:
```suggestion
	spread = FALSE
```

File: code/game/turfs/simulated/minerals.dm [hunk 14]
```diff
@@ -21,6 +21,7 @@
 	var/spreadChance = 0 //the percentual chance of an ore spreading to the neighbouring tiles
 	var/scan_state = "" //Holder for the image we display when we're pinged by a mining scanner
 	var/defer_change = FALSE
+	var/hardness = 1 //how hard the material is, we'll have to have more powerful stuff if we want to blast harder materials.
```
Comment (Bibby0110) [inline code/game/turfs/simulated/minerals.dm:24]: use defines

**Review by Bibby0110 — DISMISSED** [review]

File: code/modules/mining/ores_coins.dm (hunk 1 above)
Comment (ToasterBiome) [reply code/modules/mining/ores_coins.dm:301]: what do u mean no

File: code/modules/mining/lavaland/world_anvil.dm (hunk 10 above)
Comment (ToasterBiome) [reply code/modules/mining/lavaland/world_anvil.dm:27]: it has ..()

File: code/modules/mining/lavaland/world_anvil.dm (hunk 6 above)
Comment (ToasterBiome) [reply code/modules/mining/lavaland/world_anvil.dm]: havent gotten around to gun

File: code/game/turfs/simulated/minerals.dm (hunk 13 above)
Comment (ToasterBiome) [reply code/game/turfs/simulated/minerals.dm:799]: other minerals use numbers

File: code/game/turfs/simulated/minerals.dm (hunk 14 above)
Comment (ToasterBiome) [reply code/game/turfs/simulated/minerals.dm:24]: **#define ROCK_HARD 0
#define ROCK_HARDER 1
#define ROCK_HARDERER 2
#define ROCK_HARDERERER 3**

Comment (Bibby0110) [reply code/game/turfs/simulated/minerals.dm:24]: beautiful

File: code/modules/mining/lavaland/world_anvil.dm (hunk 10 above)
Comment (Bibby0110) [reply code/modules/mining/lavaland/world_anvil.dm:27]: where?

File: code/modules/mining/ores_coins.dm (hunk 1 above)
Comment (Bibby0110) [reply code/modules/mining/ores_coins.dm:301]: lowercase and seperated by underscore
magmite_ore

File: code/modules/mining/lavaland/world_anvil.dm (hunk 10 above)
Comment (ToasterBiome) [reply code/modules/mining/lavaland/world_anvil.dm:27]: line 24

File: code/modules/mining/ores_coins.dm (hunk 1 above)
Comment (ToasterBiome) [reply code/modules/mining/ores_coins.dm:301]: this matches the other ore icons, so no

**Comment by bruhlookatthisdood** [comment]
maybe the anvil should only be usable or only appear after defeating legion?

**Comment by ToasterBiome** [comment]
> maybe the anvil should only be usable or only appear after defeating legion?

For what purpose

**Comment by bruhlookatthisdood** [comment]
> > maybe the anvil should only be usable or only appear after defeating legion?
> 
> For what purpose

For the meme.

File: yogstation/code/game/turfs/simulated/minerals.dm [hunk 15]
```diff
@@ -11,3 +11,7 @@
 	baseturfs = /turf/open/floor/plating/asteroid/basalt/lava_land_surface
 	initial_gas_mix = LAVALAND_DEFAULT_ATMOS
 	defer_change = 1
+
+/turf/closed/mineral/dilithium/volcanic/hard
+	smooth_icon = 'icons/turf/smoothrocks_hard.dmi'
+	hardness = 1
```
Comment (SomeguyManperson) [inline yogstation/code/game/turfs/simulated/minerals.dm]: ??

Comment (ToasterBiome) [reply yogstation/code/game/turfs/simulated/minerals.dm]: what does this mean

**Pull request merged**